mqtt/mqtt-broker/tests/common.rs

```diff
@@ -0,0 +1,179 @@
+use std::{
+    sync::atomic::{AtomicU32, Ordering},
+    time::Duration,
+};
+
+use futures_util::{FutureExt, StreamExt};
+use lazy_static::lazy_static;
+use tokio::net::ToSocketAddrs;
+use tokio::sync::{
+    mpsc::{self, Receiver},
+    oneshot::{self, Sender},
+};
+use tokio::task::{JoinError, JoinHandle};
+
+use mqtt3::{
+    proto::{Publication, SubscribeTo},
+    Client, Event, PublishError, PublishHandle, ShutdownError, ShutdownHandle,
+    UpdateSubscriptionError, UpdateSubscriptionHandle,
+};
+use mqtt_broker::{Authenticator, Authorizer, Broker, BrokerState, Error, Server};
+
+/// A wrapper on the [`mqtt3::Client`] to help simplify client event loop management.
+#[derive(Debug)]
+pub struct TestClient {
+    publish_handle: PublishHandle,
+    subscription_handle: UpdateSubscriptionHandle,
+    shutdown_handle: ShutdownHandle,
+    pub events_receiver: Receiver<Event>,
+    task: JoinHandle<()>,
+}
+
+impl TestClient {
+    pub async fn publish(&mut self, publication: Publication) -> Result<(), PublishError> {
+        self.publish_handle.publish(publication).await
+    }
+
+    pub async fn subscribe(
+        &mut self,
+        subscribe_to: SubscribeTo,
+    ) -> Result<(), UpdateSubscriptionError> {
+        self.subscription_handle.subscribe(subscribe_to).await
+    }
+
+    pub async fn shutdown(&mut self) -> Result<(), ShutdownError> {
+        self.shutdown_handle.shutdown().await
+    }
+
+    pub async fn join(self) -> Result<(), JoinError> {
+        self.task.await
+    }
+}
+
+pub struct TestClientBuilder<T>
+where
+    T: ToSocketAddrs + Clone + Send + Sync + Unpin + 'static,
+{
+    address: T,
+    client_id: Option<String>,
+    username: Option<String>,
+    password: Option<String>,
+    will: Option<Publication>,
+}
+
+#[allow(dead_code)]
+impl<T> TestClientBuilder<T>
+where
+    T: ToSocketAddrs + Clone + Send + Sync + Unpin + 'static,
+{
+    pub fn new(address: T) -> Self {
+        Self {
+            address,
+            client_id: None,
+            username: None,
+            password: None,
+            will: None,
+        }
+    }
+
+    pub fn address(mut self, address: T) -> Self {
```

Comment:
Do you still need this method since you have `new(address: T)`?

Reply:
No need - removed.